### PR TITLE
Mesh controler

### PR DESCRIPTION
**Pull request by Luckabarbeau**
# Description of the problem

- The maximal number of elements is not always taken into consideration correctly.
- We have no way to control the exact number of elements in the domain.

# Description of the solution

-Create a mesh refinement controller that controls the number of elements in the domain to be equal to the maximal number of elements.

# How Has This Been Tested?

- Multiple examples have been run with this enables.
- Creating a test with it would be very costly since multiple time steps are needed.

# Documentation

- The new parameters have been added.

### CODE REVIEW

File: source/core/mesh_controller.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+using namespace std;
+double
+MeshController::calculate_coarsening_factor(int current_number_of_cell)
+{
+  // Parameter of the controller
+  double P = 0.5;
+  double I = 0.3;
+  double D = 0.1;
```
Comment (hepap) [inline source/core/mesh_controller.cc:15]: Will these parameters always work (i.e., could it be interesting to add them in the .prm)?

File: include/core/mesh_controller.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+#ifndef lethe_mesh_controler_h
+#define lethe_mesh_controler_h
+
+#include <iostream>
+
+
+/**
+ * Controller that target a given number of elements in the mesh. This
+ * controller is use to define the coarsening factor.
```
Comment (oguevremont) [inline include/core/mesh_controller.h:29]:
```suggestion
 * controller is used to define the coarsening factor.
```

File: include/core/mesh_controller.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+class MeshController
+{
+public:
+  // Constructor with the number of elements
+  MeshController(int target_number_of_elements)
+    : target_number_of_elements(target_number_of_elements)
+  {}
+  virtual ~MeshController()
+  {}
+  // Function that return the coarsening factor.
```
Comment (oguevremont) [inline include/core/mesh_controller.h:41]:
```suggestion
/**
* @brief Returns the coarsening factor.
*/
```

File: include/core/mesh_controller.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Controller that target a given number of elements in the mesh. This
+ * controller is use to define the coarsening factor.
+ */
+
+class MeshController
+{
+public:
+  // Constructor with the number of elements
```
Comment (oguevremont) [inline include/core/mesh_controller.h:35]:
```suggestion
/**
* @brief Constructor
* @param target_number_of_elements
*/
```

File: source/core/mesh_controller.cc [hunk 5]
```diff
@@ -0,0 +1,51 @@
+//
+// Created by luckab on 12/07/23.
+//
+
```
Comment (oguevremont) [inline source/core/mesh_controller.cc:4]:
```suggestion
```

File: source/core/mesh_controller.cc [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#include <core/mesh_controller.h>
+
+
+using namespace std;
+double
+MeshController::calculate_coarsening_factor(int current_number_of_cell)
```
Comment (oguevremont) [inline source/core/mesh_controller.cc:10]: To change with refactor
```suggestion
MeshController::calculate_coarsening_factor(int current_number_of_cells)
```

File: source/solvers/navier_stokes_base.cc [hunk 7]
```diff
@@ -899,12 +902,24 @@ NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
...
   std::vector<bool> global_coarsen_flags(dim * tria.n_active_cells(), false);
 
-  bool first_variable(true);
+  bool   first_variable(true);
+  double coarsening_factor = mesh_controller.calculate_coarsening_factor(
+    this->triangulation->n_global_active_cells());
+
+  unsigned int maximal_number_of_element =
+    this->simulation_parameters.mesh_adaptation.maximum_number_elements;
+  // Override the maximal number of element if the controller is use.
```
Comment (oguevremont) [inline source/solvers/navier_stokes_base.cc:911]:
```suggestion
  // Override the maximal number of element if the controller is used.
```

File: source/solvers/navier_stokes_base.cc [hunk 8]
```diff
@@ -899,12 +902,24 @@ NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
...
+  if (this->simulation_parameters.mesh_adaptation.enable_mesh_controller)
+    maximal_number_of_element = INT_MAX;
```
Comment (oguevremont) [inline source/solvers/navier_stokes_base.cc:913]:
```suggestion
    maximal_number_of_element = std::numeric_limits<int>::max();
```

**Comment by oguevremont** [comment]
An idea for the future, but if it becomes necessary we could add P, I and D parameters, but given them default values like the one that are currently hard-coded.

File: source/solvers/navier_stokes_base.cc [hunk 9]
```diff
@@ -899,12 +902,24 @@ NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
   std::vector<bool> global_refine_flags(dim * tria.n_active_cells(), false);
   std::vector<bool> global_coarsen_flags(dim * tria.n_active_cells(), false);
 
-  bool first_variable(true);
+  bool   first_variable(true);
+  double coarsening_factor = mesh_controller.calculate_coarsening_factor(
+    this->triangulation->n_global_active_cells());
+
+  unsigned int maximal_number_of_element =
```
Comment (hepap) [inline source/solvers/navier_stokes_base.cc:909]:
```suggestion
  unsigned int maximal_number_of_elements =
```

File: doc/source/parameters/cfd/mesh_adaptation_control.rst [hunk 10]
```diff
@@ -86,8 +89,16 @@ This subsection controls the mesh adaptation method, with default values given b
 
 * The maximum number of elements in the entire domain can be controlled with the ``max number elements`` parameter.
 
+* The boolean parameter ``enable mesh refinement controller`` activate a controller that overrides the value of the of ``fraction coarsening`` parameter. If it is activated, the controller will try to maintain the total number of elements in the domain equal to the value of ``max number elements`` parameter.
+
+.. note:: 
+    If the ``fraction refinement`` parameter is too high, the controller may not be able to maintain the number of elements constant. If ``fraction type = number``, the maximal ``fraction refinement`` that is stable in 3D is 0.125. In 2D, it is 0.25.
```
[the remark this answers is not in the document]
Comment (oguevremont) [reply doc/source/parameters/cfd/mesh_adaptation_control.rst:95]: Alright, I didn't understand it well before

**Review by oguevremont — APPROVED** [review]

File: doc/source/parameters/cfd/mesh_adaptation_control.rst [hunk 11]
```diff
@@ -7,38 +7,41 @@ This subsection controls the mesh adaptation method, with default values given b
...
 
 	  # Fraction of refined elements
 	  # For multi-variables refinement, separate the different fractions with a comma
-	  set fraction refinement      = 0.1
+	  set fraction refinement               = 0.1
 
 	  # How the fraction of refinement/coarsening are interpreted
 	  # Choices are number or fraction 
-	  set fraction type            = number
+	  set fraction type                     = number
```
Comment (blaisb) [inline doc/source/parameters/cfd/mesh_adaptation_control.rst:35]: Are those characters tabs? We moved away from using tabs in the .rst file it breaks on some renderers. Can you double check that there are no tabs?

File: doc/source/parameters/cfd/mesh_adaptation_control.rst [hunk 12]
```diff
@@ -86,8 +89,16 @@ This subsection controls the mesh adaptation method, with default values given b
 
 * The maximum number of elements in the entire domain can be controlled with the ``max number elements`` parameter.
 
+* The boolean parameter ``enable mesh refinement controller`` activates a controller that overrides the value of the of ``fraction coarsening`` parameter. If activated, the controller will try to maintain the total number of elements in the domain equal to the value of ``max number elements`` parameter. The control is done using a PID controller.
```
Comment (blaisb) [inline doc/source/parameters/cfd/mesh_adaptation_control.rst:92]:
```suggestion
* The boolean parameter ``enable mesh refinement controller`` activates a controller that overrides the value of the of ``fraction coarsening`` parameter. If activated, the controller will try to maintain the total number of elements in the domain equal to the value of ``max number elements`` parameter. The control is done using a PID controller.
```
Is it a PID or a PI?

File: include/core/mesh_controller.h [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+* it, and/or modify it under the terms of the GNU Lesser General
+* Public License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+* The full text of the license can be found in the file LICENSE at
+* the top level of the Lethe distribution.
+*
+* ---------------------------------------------------------------------
+
+*
+* Author: Lucka, Polytechnique Montreal, 2023 -
```
Comment (blaisb) [inline include/core/mesh_controller.h:17]:
```suggestion
```
We moved away from using author names in the files.

File: include/core/mesh_controller.h [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+#ifndef lethe_mesh_controler_h
+#define lethe_mesh_controler_h
+
+#include <iostream>
+
+
+/**
+ * Controller that target a given number of elements in the mesh. This
+ * controller is used to define the coarsening factor.
+ */
```
Comment (blaisb) [inline include/core/mesh_controller.h:30]: use @class when writing the brief for the class

File: include/core/mesh_controller.h [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+class MeshController
+{
+public:
+  /**
+   * @brief Constructor
+   * @param target_number_of_elements
```
Comment (blaisb) [inline include/core/mesh_controller.h:37]: More verbose comments please

File: include/core/mesh_controller.h [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  MeshController(int target_number_of_elements)
```
Comment (blaisb) [inline include/core/mesh_controller.h:39]:
```suggestion
  MeshController(const unsigned int target_number_of_elements)
```

File: include/core/mesh_controller.h [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    : target_number_of_elements(target_number_of_elements)
+  {}
+  virtual ~MeshController()
+  {}
+  /**
+   * @brief Returns the coarsening factor.
+   */
+  double
+  calculate_coarsening_factor(int current_number_of_elements);
```
Comment (blaisb) [inline include/core/mesh_controller.h:48]:
```suggestion
  calculate_coarsening_factor(const unsigned int current_number_of_elements);
```

File: include/core/mesh_controller.h [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+private:
+  const int target_number_of_elements;
+  int       previous_number_of_elements;
```
Comment (blaisb) [inline include/core/mesh_controller.h:52]:
```suggestion
  const unsigned int target_number_of_elements;
  unsigned int       previous_number_of_elements;
```

File: source/core/CMakeLists.txt [hunk 19]
```diff
@@ -28,6 +28,7 @@ add_library(lethe-core
   thermal_expansion_model.cc
   tracer_diffusivity_model.cc
   utilities.cc
+  mesh_controller.cc
```
Comment (blaisb) [inline source/core/CMakeLists.txt:31]: Please put file in alphabetical order

File: source/core/CMakeLists.txt [hunk 20]
```diff
@@ -69,7 +70,8 @@ add_library(lethe-core
   ../../include/core/time_integration_utilities.h
   ../../include/core/tracer_diffusivity_model.h
   ../../include/core/exceptions.h
-  ../../include/core/utilities.h)
+  ../../include/core/utilities.h
+  ../../include/core/mesh_controller.h)
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/core/CMakeLists.txt:74]: Not could, but should

File: source/core/mesh_controller.cc [hunk 21]
```diff
@@ -0,0 +1,47 @@
+#include <core/mesh_controller.h>
+
+
+using namespace std;
+double
+MeshController::calculate_coarsening_factor(int current_number_of_elements)
+{
```
Comment (blaisb) [inline source/core/mesh_controller.cc:7]: Can you write a unit test for this class

File: source/core/mesh_controller.cc [hunk 22]
```diff
@@ -0,0 +1,47 @@
+#include <core/mesh_controller.h>
+
+
+using namespace std;
```
Comment (blaisb) [inline source/core/mesh_controller.cc:4]:
```suggestion
```
Let's avoid these namespace. We don't have that in the library (or we should not)

File: source/core/mesh_controller.cc [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+MeshController::calculate_coarsening_factor(int current_number_of_elements)
+{
+  // Parameters of the controller
+  double P = 0.5;
+  double I = 0.3;
+  double D = 0.1;
+
+  // Evaluation of the error used to control the mesh refinement.
+  double error = (target_number_of_elements - current_number_of_elements) /
+                 (double)target_number_of_elements;
```
Comment (blaisb) [inline source/core/mesh_controller.cc:15]:
```suggestion
                 (double)target_number_of_elements;
```
Should be a static_cast<double>

File: source/core/mesh_controller.cc [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+  double previous_error =
+    (target_number_of_elements - previous_number_of_elements) /
+    (double)target_number_of_elements;
```
Comment (blaisb) [inline source/core/mesh_controller.cc:18]: static cast

File: source/core/mesh_controller.cc [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+  // Saturate the coarsening parameter.
+  if (coarsening_fraction_controled < 0.0)
+    {
+      // Stop the integration if we are at the saturation point.
+      coarsening_fraction_controled = 0;
+      if (error > 0.0)
+        {
+          previous_mesh_control_error = previous_mesh_control_error - error;
+        }
```
Comment (blaisb) [inline source/core/mesh_controller.cc:34]:
```suggestion
      if (error > 0.0)
          previous_mesh_control_error = previous_mesh_control_error - error;
```

File: source/core/mesh_controller.cc [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  if (coarsening_fraction_controled > 1.0)
+    {
+      coarsening_fraction_controled = 1;
+      // Stop the integration if we are at the saturation point.
+      if (error < 1.0)
+        {
```
Comment (blaisb) [inline source/core/mesh_controller.cc:41]:
```suggestion
```

File: source/core/mesh_controller.cc [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+          previous_mesh_control_error = previous_mesh_control_error - error;
+        }
```
Comment (blaisb) [inline source/core/mesh_controller.cc:43]:
```suggestion
```

File: source/core/parameters.cc [hunk 28]
```diff
@@ -1926,6 +1926,13 @@ namespace Parameters
                         "1",
                         Patterns::Integer(),
                         "Frequency of the mesh refinement");
+
+      prm.declare_entry(
+        "enable mesh refinement controller",
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/core/parameters.cc:1931]: I concur. I would also remove the enable. It's a booleann, so if it's true it's enabled

**Review by blaisb — CHANGES_REQUESTED** [review]
Some changes there and there. It would be great to have a unit test for the feature though.

File: include/core/mesh_controller.h (hunk 13 above)
Comment (blaisb) [inline include/core/mesh_controller.h:17]: Stil needs to be removed

File: source/core/mesh_controller.cc [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+MeshController::calculate_coarsening_factor(
+  unsigned int current_number_of_elements)
+{
+  // Parameters of the controller
+  double P = 0.5;
+  double I = 0.3;
+  double D = 0.1;
+
+  // Evaluation of the error used to control the mesh refinement.
+  double error = static_cast<signed int>(target_number_of_elements -
```
Comment (blaisb) [inline source/core/mesh_controller.cc:13]:
```suggestion
  double error = static_cast<double>(target_number_of_elements -
```

File: source/core/mesh_controller.cc [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+                                         current_number_of_elements) /
+                 static_cast<double>(target_number_of_elements);
+  double previous_error = static_cast<signed int>(target_number_of_elements -
```
Comment (blaisb) [inline source/core/mesh_controller.cc:16]:
```suggestion
  double previous_error = static_cast<double>(target_number_of_elements -
```

File: source/solvers/navier_stokes_base.cc [hunk 31]
```diff
@@ -899,12 +901,27 @@ NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
   std::vector<bool> global_refine_flags(dim * tria.n_active_cells(), false);
   std::vector<bool> global_coarsen_flags(dim * tria.n_active_cells(), false);
 
-  bool first_variable(true);
+  bool   first_variable(true);
+  double coarsening_factor = mesh_controller.calculate_coarsening_factor(
```
Comment (blaisb) [inline source/solvers/navier_stokes_base.cc:905]:
```suggestion
  const double coarsening_factor = mesh_controller.calculate_coarsening_factor(
```

if it's not modified afterwards

File: source/solvers/navier_stokes_base.cc [hunk 32]
```diff
@@ -899,12 +901,27 @@ NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
...
+    this->triangulation->n_global_active_cells());
+
+  unsigned int maximal_number_of_elements =
```
Comment (blaisb) [inline source/solvers/navier_stokes_base.cc:908]:
```suggestion
  const unsigned int maximal_number_of_elements =
```
If its not modified afterwards

**Review by blaisb — APPROVED** [review]
Just 1-2 Small comments left to address

File: source/solvers/navier_stokes_base.cc (hunk 32 above)
Comment (Luckabarbeau) [reply source/solvers/navier_stokes_base.cc:908]: this is not a const

**Comment by blaisb** [comment]
@Luckabarbeau Awesome, will merge after CI

**Comment by blaisb** [comment]
@Luckabarbeau it seems your test is not portable?

**Pull request merged**